### ISSUE TITLE
[Paper Note] Linear Transformers Are Secretly Fast Weight Programmers, Imanol Schlag+, arXiv'21, 2021.02

### ISSUE THREAD

**Issue by AkihikoWatanabe** [AkihikoWatanabe/paper_notes#5294]
# URL
- https://arxiv.org/abs/2102.11174
# Authors
  - Imanol Schlag
  - Kazuki Irie
  - Jürgen Schmidhuber
# Abstract
  - We show the formal equivalence of linearised self-attention mechanisms and fast weight controllers from the early '90s, where a ``slow" neural net learns by gradient descent to program the ``fast weights" of another net through sequences of elementary programming instructions which are additive outer products of self-invented activation patterns (today called keys and values). Such Fast Weight Programmers (FWPs) learn to manipulate the contents of a finite memory and dynamically interact with it. We infer a memory capacity limitation of recent linearised softmax attention variants, and replace the purely additive outer products by a delta rule-like programming instruction, such that the FWP can more easily learn to correct the current mapping from keys to values. The FWP also learns to compute dynamically changing learning rates. We also propose a new kernel function to linearise attention which balances simplicity and effectiveness. We conduct experiments on synthetic retrieval problems as well as standard machine translation and language modelling tasks which demonstrate the benefits of our methods.
# Translation (by gpt-5-nano)
- 私たちは、90年代初頭の線形化された自己注意機構とファストウェイト・コントローラの形式的等価性を示す。ここで「遅い」とされるニューラルネットは勾配降下法によって学習し、別のネットの「ファストウェイト」をプログラムする。その一連の基本的なプログラミング命令は、自己発明の活性化パターンの加法的外積に基づくものである。こうしたファストウェイト・プログラマ（FWP）は、有限のメモリの内容を操作し、それと動的に相互作用することを学習する。我々は、最近の線形化されたソフトマックス注意の変種にはメモリ容量の限界があると推定し、純粋に加法的な外積をデルタ規則に類するプログラミング命令へ置換することで、FWP がキーから値への現在の写像をより容易に訂正できるようにする。FWP はまた、動的に変化する学習率を計算することも学習する。さらに、注意を線形化する新しいカーネル関数を提案し、それが単純さと有効性のバランスを取る。私たちは、合成的なリトリーバル問題だけでなく、標準的な機械翻訳および言語モデリングのタスクについて実験を行い、提案手法の利点を示す。
# Summary (by gpt-4o-mini)
- 線形化された自己注意機構とファストウェイト・コントローラの等価性を示し、遅いニューラルネットがファストウェイトをプログラムする方法を探る。FWPは有限メモリの操作を学習し、注意機構のメモリ容量限界を改善するために、加法的外積を用いたプログラミング命令を導入。動的学習率の計算も学習し、新しいカーネル関数を提案。合成リトリーバル問題や機械翻訳、言語モデリングタスクにおける利点を実験で示した。

**Comment by AkihikoWatanabe** [comment]
Linear TransformerにおけるKV^Tを加算する固定された状態更新ではなく、差分だけを更新するDelta Ruleによって更新する Delta Networkを提案。

（まだ全然読めていないのでしっかり読みたい）

**Comment by AkihikoWatanabe** [comment]
DeltaNet図解:
https://x.com/nrehiew_/status/2079945408609239354?s=46&t=Y6UuIHB0Lv0IpmFAjlc2-Q
